### DISCUSSION TITLE
Type action in xstate

### DISCUSSION THREAD

**Discussion by Djordjenp** [statelyai/xstate#3494] [Q&A] [answered]
Hey I created my machine using TS but I get error for actions "Property 'email' does not exist on type '{ username: string; position: string; } | { phone: string; email: string; }'.   Property 'email' does not exist on type '{ username: string; position: string; }'."

Here is my machine 
```js
import {assign, createMachine} from "xstate";

interface Context {
    username: string,
    position: string,
    phone: string,
    email: string,
    dateOfBirth: string,
    nationalId: string
}

type MultiStepFormMachineEvent =
     {
    type: 'ENTER_USERNAME';
    payload: { username: string, position: string };
}

    | {
    type: 'ENTER_CONTACT',
    payload: {phone: string, email: string}
}



const registrationFormMachine = createMachine<Context, MultiStepFormMachineEvent>({
    id: 'machine',
    initial: 'enteringUsername',
    context: {
        username: '',
        position: '',
        phone: '',
        email: '',
        dateOfBirth: '',
        nationalId: ''
    },
    states: {
        enteringUsername: {
            on: {
                ENTER_USERNAME: {
                    target: 'enteringContact',
                    actions: ['assignUsername', 'assignPosition']
                }
            }
        },
        enteringContact: {
            on: {
                ENTER_CONTACT: {
                    target: 'enteringID',
                    actions: ['assignPhone', 'assignEmail']
                }
            }
        },
        enteringID: {},
        confirming: {}
    }
}, {
    actions: {
        assignUsername: assign({
            username: (_, event) => event.payload.username
        }),
        assignPosition: assign({
            position: (_, event) => event.payload.position
        }),

        assignPhone: assign({
            phone:(_, event) => event.payload.phone
        }),

        assignEmail: assign({
            email:(_, event) => event.payload.email
        }),
    }
})

export default registrationFormMachine;


```

**Comment by christoph-fricke** [comment, marked as answer]
Hi,

when not using [Typegen](https://xstate.js.org/docs/guides/typescript.html#typegen), the TS inference in actions is limited to the union of all possible events. To solve it, you can do something like this:

```typescript
assignEmail: assign({
  email:(context, event) => event.type === "ENTER_CONTACT" ? event.payload.email : context.email
}),
```

However, I would suggest to give Typegen a try, since it improves the developer experience drastically.

**Reply by geekox86** [reply]
Thx a lot for your time and knowledge.
I think xstate needs a simple event-based choose-like action creator built-in to simplify this common use case. Creating multiple actions reduces reusability and increases verbosity and pain of naming actions.

Also, it would even be easier if on/after/entry/exit/etc. actions can have a cond/conds as I see it much better place than placing them inside the actions themselves because conditions tend to be related to the place where the actions are fired from rather than the actions themselves. This will result in much flexibility and simpler plain actions free of logic. The machine definition will be more declarative too.

What do you think?

**Reply by Andarist** [reply]
> I think xstate needs a simple event-based choose-like action creator built-in to simplify this common use case. Creating multiple actions reduces reusability and increases verbosity and pain of naming actions.

Could you give some examples here? I'm not sure if I fully understand what you mean. A simple comparison of the current state of things with your ideal API~ would help me out here.

> Also, it would even be easier if on/after/entry/exit/etc. actions can have a cond/conds as I see it much better place than placing them inside the actions themselves because conditions tend to be related to the place where the actions are fired from rather than the actions themselves. This will result in much flexibility and simpler plain actions free of logic. The machine definition will be more declarative too.

This is an interesting idea. Maybe it's worth creating an RFC about it.

**Reply by geekox86** [reply]
I will try to submit an RFC and explain the use cases better there.

As for the suggestion of an event-based choose-like action creator: instead of a generic `cond` in `choose`, I suggest to have another action creator that accepts an event type name (or list of event type names) which will automatically narrow down the events in the `actions` section of the action creator.
